extensions/amp-viewer-integration/0.1/examples/amp-viewer-host.js

```diff
@@ -0,0 +1,166 @@
+
+initLogConstructor();
+const CHANNEL_OPEN_MSG = 'channelOpen';
+
+/**
+ * @fileoverview This is an example of how the viewer host can be implemented
+ * for communication with the AMP docs.
+ */
+export class AmpViewerHost {
+
+  /**
+   * @param {!Window} win
+   * @param {!HTMLIFrameElement} ampIframe
+   * @param {string} frameOrigin
+   * @param {function(string, *, boolean):(!Promise<*>|undefined)} messageHandler
+   * @param {string=} opt_logsId For dev logs so you know what ampdoc you're
+   * @param {boolean=} opt_isWebview Should viewer initiate handshake w/ polling
+   * @param {boolean=} opt_isHandshakePoll
+   * looking at.
+   */
+  constructor(win, ampIframe, frameOrigin, messageHandler, opt_logsId,
+      opt_isWebview, opt_isHandshakePoll) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @private {!HTMLIFrameElement} */
+    this.ampIframe_ = ampIframe;
+    /** @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.messageHandler_ = messageHandler;
+    /** @const {boolean} */
+    this.isWebview_ = !!opt_isWebview;
+    /** @const {string} */
+    this.logsId = opt_logsId;
+
+    if (this.isWebview_ || opt_isHandshakePoll) {
+      /** @private {string} */
+      this.pollingIntervalId_ = setInterval(
+        this.initiateHandshake_.bind(this, this.intervalCtr) , 1000); //poll every second
+    } else {
+      this.waitForHandshake_(frameOrigin);
+    }
+  }
+
+  /**
+   * @private
+   */
+  initiateHandshake_() {
+    this.log('initiateHandshake_');
+    if (this.ampIframe_) {
+      const channel = new MessageChannel();
+      let message = {
+        app: APP,
+        name: 'handshake-poll',
+      };
+      message = this.isWebview_ ? JSON.stringify(message) : message;
+      this.ampIframe_.contentWindow./*OK*/postMessage(
+        message, '*', [channel.port2]);
+
+      channel.port1.onmessage = function(e) {
+        const data = this.isWebview_ ? JSON.parse(e.data) : e.data;
+        if (this.isChannelOpen_(data)) {
+          this.win.clearInterval(this.pollingIntervalId_); //stop polling
+          this.log('messaging established!');
+          this.completeHandshake_(channel.port1, data.requestid);
+        } else {
+          this.messageHandler_(e);
+        }
+      }.bind(this);
+    }
+  }
+
+  /**
+   * @param {string} targetOrigin
+   * @private
+   */
+  waitForHandshake_(targetOrigin) {
+    this.log('awaitHandshake_');
+    const target = this.ampIframe_.contentWindow;
+    const listener = function(event) {
+      if (event.origin == targetOrigin &&
+              this.isChannelOpen_(event.data) &&
+              (!event.source || event.source == target)) {
+        this.log(' messaging established with ', targetOrigin);
+        unlisten();
+        const port = new WindowPortEmulator(this.win, targetOrigin, target);
+        this.completeHandshake_(port, event.data.requestid);
+      }
+    }.bind(this);
+    const unlisten = listen(this.win, 'message', listener);
```

Comment:
done. but I needed to init unlisten to null b/c the linter complained that it's never actually reassigned